## Importing libraries 

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Import CSV files

In [2]:
# defining path
path = r'C:\Users\Chris Arnold\Documents\Instacart Basket Analysis'

In [3]:
# importing products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# importing orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

## Following along with sections in 4.5

# What are Data Consistency Checks?

In [5]:
# Using df.describe()
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [6]:
#Frequency check on Order_dow column
df_ords['order_dow'].value_counts()

0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: order_dow, dtype: int64

# Mixed-Type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
# using head function to view data
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# function for checking a dataframe for mixed type column
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# changing data type for column mix to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing values

In [13]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

This tells us that column product_name is missing 16 values

In [14]:
# Creating a subset of these missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
# Checking what was created
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Addressing missing values

In [16]:
#finding the shape of the data set
df_prods.shape

(49693, 5)

In [17]:
# Creating new data set to host the cleaned data
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
# Checking for changes
df_prods_clean.shape

(49677, 5)

# Duplicates

In [19]:
#Looking for full duplicates in your dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
# Checking list created
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

Addressing Duplicates

In [21]:
# Checking the shape of df_prods
df_prods_clean.shape

(49677, 5)

In [22]:
# Setting up new set up no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
# Checking the shape of the cleaned data
df_prods_clean_no_dups.shape

(49672, 5)

## Questions on this task to complete

Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. 

In [24]:
# Using df.describe()
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The prices column most likely has something that is mis priced. It is unlikely that ther ewould be something that almost costs $100,000.00 when the mean price is 9.99 and rthe standard deviation is 453.51. The department_id, product_id and aisle_id look like they should be mostly ok with the data that is showing. 

## Check for mixed-type data in your df_ords dataframe. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [27]:
#Checking for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

This came back with no output, telling us that there is no mixed data type

## Run a check for missing values in your df_ords dataframe.

    In a markdown cell, report your findings and propose an explanation for any missing values you find.


In [28]:
# finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [35]:
# counting unique user ID numbers to see if this matches the days_since_prior_order count of missing values
print(df_ords.nunique())



order_id                  3421083
user_id                    206209
eval_set                        3
order_number                  100
order_dow                       7
order_hour_of_day              24
days_since_prior_order         31
dtype: int64


Since the number of the days_since_prior_order missing value cells (206209) equals the unique user id number (206209), the only logical explanation is that every customer has one missing value for their first order. You can not have a prior order before your first order. I imagine that the company has an automatic calculator set up to input these values for days_since_prior_order (any smart company would) and that calculator would not give a numeric representation for an argument that would not make sense logically.

## Address the missing values using an appropriate method.

    In a markdown cell, explain why you used your method of choice.

In [36]:
# Replacing all missing values in the days since prior order with 0
df_ords['days_since_prior_order'].fillna( 0 , inplace=True)

In [39]:
# Double checking that action took place
# finding missing values
df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

I carried this out because this is the only replacement value that would make sense in this situation for record keeping for the company. This would also not skew any averages because this value is true from a sense that there were no days for a prior order before the first order to take place. 

## Run a check for duplicate values in your df_ords data. Address the duplicates using an appropriate method.

      In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [40]:
#Looking for full duplicates in your dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
# Checking list created
df_ords_dups.head()

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [42]:
#double checking since there was no head, showing entire set
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates to speak of so there are none to manage. Nothing to explain.

## Exporting data sets

In [43]:
#exporting cleaned orders file 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [46]:
#exporting cleaned products file
df_prods_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))